cpp/ql/src/semmle/code/cpp/models/implementations/Strdup.qll

```diff
@@ -43,3 +43,32 @@ class StrdupFunction extends AllocationFunction, ArrayFunction, DataFlowFunction
     output.isReturnValueDeref()
   }
 }
+
+/**
+ * A `strndup` style allocation function.
+ */
+class StrndupFunction extends AllocationFunction, ArrayFunction, TaintFunction {
+  StrndupFunction() {
+    exists(string name |
+      hasGlobalOrStdName(name) and
+      (
+        // strndup(str, maxlen)
+        name = "strndup"
+      )
+    )
+  }
+
+  override predicate hasArrayInput(int bufParam) { bufParam = 0 }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) { bufParam = 0 }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // This function may do only a partial copy of the input buffer to the output
+    // buffer, so it's a taint flow.
```

Comment:
I will argue this is data flow. The data-flow relation describes where there _may_ be flow, not where there _must_ be flow. For example, there is flow in `sink(x ? source() : 0);`. An implementation of `strndup` could be written along the same lines: do a full `memcpy` if the result fits, otherwise do a partial `memcpy` and a 0-termination.

Whether something is data flow or taint is not so much a hard rule but rather a question of user expectation. For example, `strcat` is taint-only because the common case is to have a non-empty destination string. Conversely, `memcpy` with unknown size is considered data flow because the common case is to copy everything of interest.

Reply:
In that case we should change the model for `strncpy` as well, similarly if `n` is large enough it performs a straight copy.

Reply:
I agree.

Reply:
Done - this involved a small change in `TaintTrackingUtil.qll`, for data flow to a dereferenced thing.